src/plugins/s390.c

```diff
@@ -398,3 +399,353 @@ gchar* bd_s390_zfcp_sanitize_lun_input (const gchar *lun, GError **error) {
+    return TRUE;
+}
+
+/*
+ * bd_s390_zfcp_scsi_offline
+ * @devno: zfcp device number
+ * @wwpn: zfcp WWPN (World Wide Port Number)
+ * @lun: zfcp LUN (Logical Unit Number)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether a LUN was successfully removed from its associated WWPN
+ */
+gboolean bd_s390_zfcp_scsi_offline(const gchar *devno, const gchar *wwpn, const gchar *lun, GError **error) {
+    FILE *scsifd = NULL;
+    FILE *fd = NULL;
+    size_t len = 0;
+    ssize_t read, rc;
+
+    const gchar *delim = " ";
+    gchar *channel = "0";
+    gchar *devid = "0";
+    gchar *path = "/proc/scsi/scsi";
+    gchar *scsidevsysfs = "/sys/bus/scsi/devices";
+
+    gchar *line = NULL;
+    gchar *fcphbasysfs = NULL;
+    gchar *fcpwwpnsysfs = NULL;
+    gchar *fcplunsysfs = NULL;
+    gchar *hba_path = NULL;
+    gchar *wwpn_path = NULL;
+    gchar *lun_path = NULL;
+    gchar *host = NULL;
+    gchar *fcplun = NULL;
+    gchar *scsidev = NULL;
+    gchar *fcpsysfs = NULL;
+    gchar *scsidel = NULL;
+    gchar **tokens = NULL;
+
+    scsifd = fopen (path, "r");
+    if (!scsifd) {
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "Failed to open path to SCSI device: %s", path);
+        return FALSE;
+    }
+
+    while ((read = getline(&line, &len, scsifd)) != -1) {
+        if (!g_str_has_prefix(line, "Host")) {
+            continue;
+        }
+
+        /* tokenize line and assign certain values we'll need later */
+        tokens = g_strsplit (line, delim, 8);
+
+        host = tokens[1];
+        fcplun = tokens[7];
+
+        scsidev = g_strdup_printf ("%s:%s:%s:%s", &host[4], channel, devid, fcplun);
+        scsidev = g_strchomp (scsidev);
+        fcpsysfs = g_strdup_printf ("%s/%s", scsidevsysfs, scsidev);
+        fcpsysfs = g_strchomp (fcpsysfs);
+        scsidel = g_strdup_printf ("%s/%s/delete", scsidevsysfs, scsidev);
+        scsidel = g_strchomp (scsidel);
+
+        /* get HBA path value (same as device number) */
+        hba_path = g_strdup_printf ("%s/hba_id", fcpsysfs);
+        len = 0; /* should be zero, but re-set it just in case */
+        fd = fopen (hba_path, "r");
+        rc = getline (&fcphbasysfs, &len, fd);
+        if (rc == -1) {
+            g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                         "Failed to read value from %s", hba_path);
+            fclose (fd);
+            g_free (hba_path);
+            return FALSE;
+        }
+        fclose (fd);
+        g_free (hba_path);
+
+        /* get WWPN value */
+        wwpn_path = g_strdup_printf ("%s/wwpn", fcpsysfs);
+        len = 0;
+        fd = fopen (wwpn_path, "r");
+        rc = getline (&fcpwwpnsysfs, &len, fd);
+        if (rc == -1) {
+            g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                         "Failed to read value from %s\n", wwpn_path);
+            g_free (wwpn_path);
+            fclose (fd);
+            return FALSE;
```

Comment:
Again many strings need to be freed here incl. ``fcpwwpnsysfs`` (see *man getline*)